jupyter_sphinx/__init__.py

```diff
@@ -1 +1,256 @@
+)
+THEBELAB_URL_DEFAULT = "https://unpkg.com/thebelab@^0.4.0"
+
+logger = logging.getLogger(__name__)
+
+
+def builder_inited(app):
+    """
+    2 cases
+    case 1: ipywidgets 7, with require
+    case 2: ipywidgets 7, no require
+    """
+    require_url = app.config.jupyter_sphinx_require_url
+    if require_url:
+        app.add_js_file(require_url)
+        embed_url = (
+            app.config.jupyter_sphinx_embed_url
+            or ipywidgets.embed.DEFAULT_EMBED_REQUIREJS_URL
+        )
+    else:
+        embed_url = (
+            app.config.jupyter_sphinx_embed_url
+            or ipywidgets.embed.DEFAULT_EMBED_SCRIPT_URL
+        )
+    if embed_url:
+        app.add_js_file(embed_url)
+
+    # add jupyter-sphinx css
+    app.add_css_file("jupyter-sphinx.css")
+    # Check if a thebelab config was specified
+    if app.config.jupyter_sphinx_thebelab_config:
+        app.add_js_file("thebelab-helper.js")
+        app.add_css_file("thebelab.css")
+
+
+def build_finished(app, env):
+    if app.builder.format != "html":
+        return
+
+    # Copy stylesheet
+    src = os.path.join(os.path.dirname(__file__), "css")
+    dst = os.path.join(app.outdir, "_static")
+    copy_asset(src, dst)
+
+    thebe_config = app.config.jupyter_sphinx_thebelab_config
+    if not thebe_config:
+        return
+
+    # Copy all thebelab related assets
+    src = os.path.join(os.path.dirname(__file__), "thebelab")
+    dst = os.path.join(app.outdir, "_static")
+    copy_asset(src, dst)
+
+
+def _setup(app):
+    """A temporary setup function so that we can use it here and in execute.
+
+    This should be removed and converted into `setup` after a deprecation
+    cycle.
+    """
+    # Configuration
+
+    app.add_config_value(
+        "jupyter_execute_kwargs",
+        dict(timeout=-1, allow_errors=True, store_widget_state=True),
+        "env",
+    )
+    app.add_config_value("jupyter_execute_default_kernel", "python3", "env")
+    app.add_config_value(
+        "jupyter_execute_data_priority",
+        [
+            WIDGET_VIEW_MIMETYPE,
+            "application/javascript",
+            "text/html",
+            "image/svg+xml",
+            "image/png",
+            "image/jpeg",
+            "text/latex",
+            "text/plain",
+        ],
+        "env",
+    )
+
+    # ipywidgets config
+    app.add_config_value("jupyter_sphinx_require_url", REQUIRE_URL_DEFAULT, "html")
+    app.add_config_value("jupyter_sphinx_embed_url", None, "html")
+
+    # thebelab config, can be either a filename or a dict
+    app.add_config_value("jupyter_sphinx_thebelab_config", None, "html")
+
+    app.add_config_value("jupyter_sphinx_thebelab_url", THEBELAB_URL_DEFAULT, "html")
+
+    # linenos config
+    app.add_config_value("jupyter_sphinx_linenos", False, "env")
+    app.add_config_value("jupyter_sphinx_continue_linenos", False, "env")
+
+    # Used for nodes that do not need to be rendered
+    def skip(self, node):
```

Comment:
Here and later: wouldn't it better to factor the inline functions and constants out of `_setup`?

Reply:
good point - latest push has this in it